core/src/main/java/org/elasticsearch/action/search/SearchScrollAsyncAction.java

```diff
@@ -0,0 +1,196 @@
+        // protect ourselves against time going backwards
+        // negative values don't make sense and we want to be able to serialize that thing as a vLong
+        return Math.max(1, System.currentTimeMillis() - startTime);
+    }
+
+    public final void run() {
+        if (scrollId.getContext().length == 0) {
+            listener.onFailure(new SearchPhaseExecutionException("query", "no nodes to search on", ShardSearchFailure.EMPTY_ARRAY));
+            return;
+        }
+        final CountDown counter = new CountDown(scrollId.getContext().length);
+        ScrollIdForNode[] context = scrollId.getContext();
+
+        for (int i = 0; i < context.length; i++) {
+            ScrollIdForNode target = context[i];
+            DiscoveryNode node = nodes.get(target.getNode());
+            if (node != null) {
+                final int shardIndex = i;
+                InternalScrollSearchRequest internalRequest = internalScrollSearchRequest(target.getScrollId(), request);
+                SearchActionListener<T> searchActionListener = new SearchActionListener<T>(null, shardIndex) {
+
+                    @Override
+                    protected void setSearchShardTarget(T response) {
+                        // don't do this - it's part of the response...
+                        assert response.getSearchShardTarget() != null : "search shard target must not be null";
+                    }
+
+                    @Override
+                    protected void innerOnResponse(T result) {
+                        assert shardIndex == result.getShardIndex() : "shard index mismatch: "
+                            + shardIndex + " but got: " + result.getShardIndex();
+                        onFirstPhaseResult(shardIndex, result);
+                        if (counter.countDown()) {
+                            try {
+                                moveToNextPhase();
+                            } catch (Exception e) {
+                                onFailure(e);
+                            }
+                        }
+                    }
+
+                    @Override
+                    public void onFailure(Exception t) {
+                        onInitialPhaseFailure(shardIndex, counter, target.getScrollId(), t);
+                    }
+                };
+                executeInitialPhase(node, internalRequest, searchActionListener);
+            } else {
+                if (logger.isDebugEnabled()) {
+                    logger.debug("Node [{}] not available for scroll request [{}]", target.getNode(), scrollId.getSource());
+                }
+                successfulOps.decrementAndGet();
+                if (counter.countDown()) {
+                    try {
+                        moveToNextPhase();
+                    } catch (RuntimeException e) {
+                        listener.onFailure(new SearchPhaseExecutionException("query", "Fetch failed", e, ShardSearchFailure.EMPTY_ARRAY));
+                        return;
```

Comment:
useless return statement? or maybe future-proof? in the latter case we should do the same above in innerOnResponse?